From 889198b2e8bfe4f2bd0bb7500669d4f5bc115b94 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Lothar=20Wa=C3=9Fmann?= <LW@KARO-electronics.de>
Date: Tue, 26 Mar 2019 10:36:09 +0100
Subject: [PATCH 13/23] imx8mm: distinguish between watchdog and softreset

On the i.MX processors the WRSR register can tell whether the last
reset was caused by a watchdog timeout or by triggering a softreset
via the WDOG_SRS bit.
Utilize this to distinguish a WDOG reset from a SOFTRESET in the
reset_cause.
---
 arch/arm/mach-imx/cpu.c | 41 +++++++++++++++++++++--------------------
 1 file changed, 21 insertions(+), 20 deletions(-)

diff --git a/arch/arm/mach-imx/cpu.c b/arch/arm/mach-imx/cpu.c
index aa541fc600ea8fbead725a80db215a79690d2d1f..22baeab8d01213ebeab92d4f29b3c5e53b348632 100644
--- a/arch/arm/mach-imx/cpu.c
+++ b/arch/arm/mach-imx/cpu.c
@@ -10,6 +10,7 @@
 
 #include <bootm.h>
 #include <common.h>
+#include <fsl_wdog.h>
 #include <netdev.h>
 #include <linux/errno.h>
 #include <asm/io.h>
@@ -37,10 +38,12 @@
 #if defined(CONFIG_DISPLAY_CPUINFO) && !defined(CONFIG_SPL_BUILD)
 static u32 reset_cause = -1;
 
-static char *get_reset_cause(void)
+static const char *get_reset_cause(void)
 {
 	u32 cause;
 	struct src *src_regs = (struct src *)SRC_BASE_ADDR;
+	struct watchdog_regs *wdog = (struct watchdog_regs *)WDOG1_BASE_ADDR;
+	u16 wrsr = readw(&wdog->wrsr);
 
 	cause = readl(&src_regs->srsr);
 #ifndef CONFIG_ANDROID_BOOT_IMAGE
@@ -51,45 +54,43 @@ static char *get_reset_cause(void)
 #endif
 	reset_cause = cause;
 
-	switch (cause) {
-	case 0x00001:
-	case 0x00011:
+	if (cause == 1)
 		return "POR";
-	case 0x00004:
+	if (cause & 0x00004)
 		return "CSU";
-	case 0x00008:
+	if (cause & 0x00008)
 		return "IPP USER";
-	case 0x00010:
+	if (cause & 0x00010)
+		if (wrsr & WRSR_SFTW)
+			return "SOFT";
 #ifdef	CONFIG_MX7
 		return "WDOG1";
 #else
 		return "WDOG";
 #endif
-	case 0x00020:
+	if (cause & 0x00020)
 		return "JTAG HIGH-Z";
-	case 0x00040:
+	if (cause & 0x00040)
 		return "JTAG SW";
-	case 0x00080:
+	if (cause & 0x00080)
 		return "WDOG3";
 #ifdef CONFIG_MX7
-	case 0x00100:
+	if (cause & 0x00100)
 		return "WDOG4";
-	case 0x00200:
+	if (cause & 0x00200)
 		return "TEMPSENSE";
 #elif defined(CONFIG_IMX8M)
-	case 0x00100:
+	if (cause & 0x00100)
 		return "WDOG2";
-	case 0x00200:
+	if (cause & 0x00200)
 		return "TEMPSENSE";
 #else
-	case 0x00100:
+	if (cause & 0x00100)
 		return "TEMPSENSE";
-	case 0x10000:
+	if (cause & 0x10000)
 		return "WARM BOOT";
 #endif
-	default:
-		return "unknown reset";
-	}
+	return "unknown reset";
 }
 
 #ifdef CONFIG_ANDROID_BOOT_IMAGE
@@ -97,7 +98,7 @@ void get_reboot_reason(char *ret)
 {
 	struct src *src_regs = (struct src *)SRC_BASE_ADDR;
 
-	strcpy(ret, (char *)get_reset_cause());
+	strcpy(ret, get_reset_cause());
 	/* clear the srsr here, its state has been recorded in reset_cause */
 	writel(reset_cause, &src_regs->srsr);
 }
-- 
2.7.4

